File: app/src/main/java/com/aslansari/hypocoin/HypoCoinApp.java
package com.aslansari.hypocoin;

import android.app.Application;

import androidx.annotation.NonNull;

import com.aslansari.hypocoin.di.AppContainer;

import java.io.IOException;
import java.util.Properties;

import timber.log.Timber;

public class HypoCoinApp extends Application {

    private static HypoCoinApp thisInstance = null;
    public AppContainer appContainer;

    @Override
    public void onCreate() {
        super.onCreate();
        thisInstance = this;

        appContainer = new AppContainer(this);
        initializeTimber();

    }

    public static HypoCoinApp getInstance() {
        return thisInstance;
    }

    private void initializeTimber(){
        if (BuildConfig.DEBUG) {
            Timber.plant(new Timber.DebugTree() {
                // Add the line number to the tag
                @Override
                protected String createStackElementTag(@NonNull StackTraceElement element) {
                    return String.format("[Line - %s] [Method - %s] [Class - %s]",
                            element.getLineNumber(),
                            element.getMethodName(),
                            super.createStackElementTag(element));
                }
            });
        }
    }
}

File: app/src/main/java/com/aslansari/hypocoin/app/HypoCoinApp.kt
package com.aslansari.hypocoin.app

import android.app.Application
import com.aslansari.hypocoin.BuildConfig
import com.google.firebase.FirebaseApp
import dagger.hilt.android.HiltAndroidApp
import timber.log.Timber
import timber.log.Timber.DebugTree

@HiltAndroidApp
class HypoCoinApp : Application() {

    override fun onCreate() {
        super.onCreate()
        initializeTimber()
        FirebaseApp.initializeApp(this)
    }

    private fun initializeTimber() {
        if (BuildConfig.DEBUG) {
            Timber.plant(object : DebugTree() {
                // Add the line number to the tag
                override fun createStackElementTag(element: StackTraceElement): String {
                    return String.format(
                        "[Line - %s] [Method - %s] [Class - %s]",
                        element.lineNumber,
                        element.methodName,
                        super.createStackElementTag(element)
                    )
                }
            })
        }
    }
}